### PR TITLE
CDAP-1772 ETL Adapter Stages Validation Logic

### PR DESCRIPTION
**Pull request by gokulavasan**
Added validation logic to ETL Template to make sure the types of the source, transforms and sink match. Validation is done during the configureAdapter phase and thus we will catch type mismatch during adapter creation process instead of failing in runtime.
 
This work required TypeResolver which is a package private class in the Guava version used. So added TypeResolution class to circumvent that restriction. Also added a number of tests to check the validation logic.

JIRA : https://issues.cask.co/browse/CDAP-1772
Bamboo : http://builds.cask.co/browse/CDAP-DT38

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java [hunk 1]
```diff
@@ -43,11 +52,17 @@
  * @param <T> type of the configuration object
  */
 public abstract class ETLTemplate<T> extends ApplicationTemplate<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(ETLTemplate.class);
   private static final Gson GSON = new Gson();
```
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java]: This is not used anywhere?

Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java]: Ah never mind, Chrome search did not work well =(

**Comment by gokulavasan** [comment]
Build passed : http://builds.cask.co/browse/CDAP-DT38

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java [hunk 2]
```diff
@@ -159,4 +162,110 @@ public void configureAdapter(String adapterName, T config, AdapterConfigurer con
...
+      // ToResolveType is current unresolved type;
+      // Ex: Actual = String; Formal = T; ToResolve = List<T>;  ==> newType = List<String>
+      Type actualType = resTypeList.get(i - 1);
+      Type formalType = unresTypeList.get(i - 1);
+      Type toResolveType = unresTypeList.get(i);
+      try {
+        Type newType;
+        newType = (new TypeResolution()).where(formalType, actualType).resolveType(toResolveType);
+
+        // If TypeResolution was not able to resolve or if the toResolveType is a TypeVariable, then try to resolve
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java]: what's an example of not being able to resolve?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java [hunk 3]
```diff
@@ -159,4 +162,110 @@ public void configureAdapter(String adapterName, T config, AdapterConfigurer con
...
+        // using just the previous resolved type.
+        // Ex: Actual = List<String> ; Formal = List<T> ; ToResolve = T ==> newType = String which is not correct;
+        // newType should be List<String>. Hence resolve only from the previous resolved type (Actual)
+        if (newType.equals(toResolveType) || (toResolveType instanceof TypeVariable)) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java]: could do ```if (toResolveType instanceof TypeVariable) { what you have below; } else { what you have above;}``` to save one call to resolveType

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java [hunk 4]
```diff
@@ -159,4 +162,110 @@ public void configureAdapter(String adapterName, T config, AdapterConfigurer con
...
+    } else {
+      unresTypeList.add(sinkToken.resolveType(inType).getType());
+    }
+
+    // Invoke validation method with list of unresolved types
+    validateTypes(unresTypeList);
+  }
+
+  @VisibleForTesting
+  static void validateTypes(ArrayList<Type> unresTypeList) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java]: can you add some comments with an example input and an example of the resolved list it builds?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/co/cask/cdap/templates/etl/common/ValidationTest.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testSimpleInvalidMatch() throws Exception {
+    // String --> Integer
+    ArrayList<Type> typeList = Lists.newArrayList();
+    typeList.add(String.class);
+    typeList.add(Integer.class);
+    ETLTemplate.validateTypes(typeList);
+  }
+
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/co/cask/cdap/templates/etl/common/ValidationTest.java]: can you add a more complicated invalid test case?

**Comment by albertshau** [comment]
couple comments

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/co/cask/cdap/templates/etl/common/ValidationTest.java (hunk 5 above)
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/co/cask/cdap/templates/etl/common/ValidationTest.java]: Added

**Comment by albertshau** [comment]
lgtm

**Comment by gokulavasan** [comment]
Build passed : http://builds.cask.co/browse/CDAP-DT38-11

**Pull request merged**